### ISSUE TITLE
Consider changing the behavior of text inputs in Blazor to treat empty string as null

### ISSUE THREAD

**Issue by Eirenarch** [dotnet/aspnetcore#51639]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Currently if the developer defines a text property as nullable (optional/not required) and the user touches the text input this triggers the validation and the user cannot remove the value. It is also impossible to remove the value in edit scenarios. This means that the developer must provide custom code to change every value from empty string to null. This is especially problematic when there is a min length validation.

### Describe the solution you'd like

Consider changing the behavior of text inputs in Blazor to automatically convert empty strings to null

While this suggestion is a breaking change I feel like it will be beneficial because it is hard to imagine many people want the empty string behavior. I also believe that ASP.NET MVC textboxes treat empty strings as null so there will be some consistency. Alternatively there might be a switch somewhere to turn on the null behavior. Another alternative is to provide another control with the new behavior. While creating such a control is not hard it will be good to not have to write it for every project.

### Additional context

This issue has been raised before but closed without consideration - https://github.com/dotnet/aspnetcore/issues/40785

**Comment by thhunt** [comment]
Yes, this please. The same sort of functionality exists with ASP .Net Core / MVC where POST back to the controller from a form maps empty string to null for nullable properties. This would prevent a lot of manual code on either model properties or component event handlers.

**Comment by kgar** [comment]
Adding to the discussion:

I'm working on a Blazor application where the database expects null or a well-formed phone number string. The db column has a check constraint on it. This is an established database over which I have no control.

Using a Blazor EditForm and binding to a nullable string results in an empty string when I clear out a phone number, for example. This breaks validation, and if validation is bypassed, it triggers a check constraint error on attempting to save the data.

The workaround I'm using is to update the relevant properties to convert empty string to null in the setter.

That Blazor project is going to scale quite a bit, so I can expect to see this workaround propagated across hundreds of models.

**Comment by Prosolve** [comment]
Just to add a further comment in an attempt to raise the priority on this. Fairly new to Blazor and initially I assumed I must be doing something wrong. The fact null [valid] and empty string [invalid] act differently for basic validation like [Phone] and [Email] seems a major bug to me. FWIW.

**Comment by thhunt** [comment]
In the FWIW category, I've gotten around this by overriding the `SaveChanges` methods in my `DbContext` so that before doing any save operations it will call a method I wrote to trim all string fields, and if the field is nullable and `string.Empty` then it will set the value to `null`. You can also put more granular control on this by adding custom attributes to the property to indicate how to handle the string value (trim action, nullability, etc.).
Doesn't help the in between of the frontend to the save but will at least help gatekeep it before hitting the database.

**Comment by Prosolve** [comment]
> In the FWIW category, I've gotten around this by overriding the `SaveChanges` methods in my `DbContext` ... <

Ah for me the error shows before trying to persist the data, which I want i.e. I can't click the Submit button. So I just created very simple validator classes for each of the main default ones such as [Phone] [Email] etc. as below then just use [OptionalPhone] decorator in my DTOs

public class OptionalPhone : ValidationAttribute
{
    public override bool IsValid(object? value)
    {
        var str = value as string;
        if (string.IsNullOrWhiteSpace(str))
            return true;

        var origAttrib = new PhoneAttribute();
        return origAttrib.IsValid(str);
    }
}
